### PR TITLE
[SPARK-8658] [SQL] AttributeReference's equals method compares all the members

### PR DESCRIPTION
**Pull request by gatorsmile**
This fix is to change the equals method to check all of the specified fields for equality of AttributeReference.

### CODE REVIEW

**Comment by JoshRosen** [comment]
Can you add a regression test?

**Comment by JoshRosen** [comment]
Actually, maybe we don't need one.

**Comment by JoshRosen** [comment]
Jenkins, this is ok to test.

**Comment by gatorsmile** [comment]
My code change expose a new defect: 
  Both rollup and cube are not working correctly no matter whether the build include my changes or not. 

Without my changes, the outputs of rollup query are
[3,2,-1]
[3,null,null]
[6,4,-2]
[6,null,null]
[null,null,null]

However, the expected results should be like
[3,2,-1]
[3,null,-1]
[6,4,-2]
[6,null,-2]
[null,null,-3]

I need more time to find the root cause why rollup and cube do not work. The current test cases in HiveDataFrameAnalyticsSuite hide the errors because both the results of both SQL and DataFrame are wrong, although the results are the same. 

Thanks, 

Xiao Li

**Comment by cloud-fan** [comment]
Hi @gatorsmile , thanks for working on it!
Do you mind opening a new JIRA to describe the rollup and cube bug you discovered? I think we should open another PR to fix that bug.

**Comment by gatorsmile** [comment]
Hi, @cloud-fan 

Sure. Will do. I am trying to see if I can easily fix it. Anyway, I will open a JIRA tonight.

Thanks, 

Xiao Li

**Comment by gatorsmile** [comment]
The JIRA is opened:
https://issues.apache.org/jira/browse/SPARK-11275 
I will continue the investigation on this JIRA issue.

**Comment by gatorsmile** [comment]
@JoshRosen @cloud-fan 
I submitted a pull request for JIRA Spark-11275: https://github.com/apache/spark/pull/9419 
Hopefully, after finishing the problem, this one can pass all the tests. Thanks!

**Comment by marmbrus** [comment]
Thanks, I'm going to merge this to master and 1.6.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala [hunk 1]
```diff
@@ -194,7 +194,9 @@ case class AttributeReference(
   def sameRef(other: AttributeReference): Boolean = this.exprId == other.exprId
 
   override def equals(other: Any): Boolean = other match {
-    case ar: AttributeReference => name == ar.name && exprId == ar.exprId && dataType == ar.dataType
+    case ar: AttributeReference =>
+      name == ar.name && dataType == ar.dataType && nullable == ar.nullable &&
+        metadata == ar.metadata && exprId == ar.exprId && qualifiers == ar.qualifiers
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala]: @gatorsmile , can you send a follow-up PR to also update the `hashCode` according to this? thanks!

Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala]: sure, will do it tonight.

Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala]: Good catch!

**Pull request closed without merging**